pkg/controller/submariner/submariner_controller.go

```diff
@@ -200,55 +204,64 @@ func (r *ReconcileSubmariner) reconcileEngineDaemonSet(instance *submarinerv1alp
 }
 
 func (r *ReconcileSubmariner) reconcileRouteagentDaemonSet(instance *submarinerv1alpha1.Submariner, reqLogger logr.Logger) (*appsv1.DaemonSet, error) {
-	return r.reconcileDaemonSet(instance, newRouteAgentDaemonSet(instance), reqLogger)
+	return helpers.ReconcileDaemonSet(instance, newRouteAgentDaemonSet(instance), reqLogger, r.client, r.scheme)
 }
 
 func (r *ReconcileSubmariner) reconcileGlobalnetDaemonSet(instance *submarinerv1alpha1.Submariner, reqLogger logr.Logger) (*appsv1.DaemonSet, error) {
-	return r.reconcileDaemonSet(instance, newGlobalnetDaemonSet(instance), reqLogger)
+	return helpers.ReconcileDaemonSet(instance, newGlobalnetDaemonSet(instance), reqLogger, r.client, r.scheme)
 }
 
-func (r *ReconcileSubmariner) reconcileDaemonSet(owner metav1.Object, daemonSet *appsv1.DaemonSet, reqLogger logr.Logger) (*appsv1.DaemonSet, error) {
-	var err error
-
-	// Set the owner and controller
-	if err = controllerutil.SetControllerReference(owner, daemonSet, r.scheme); err != nil {
-		return nil, err
-	}
-
-	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		toUpdate := &appsv1.DaemonSet{ObjectMeta: metav1.ObjectMeta{
-			Name:      daemonSet.Name,
-			Namespace: daemonSet.Namespace,
-			Labels:    map[string]string{},
-		}}
-
-		result, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, toUpdate, func() error {
-			toUpdate.Spec = daemonSet.Spec
-			for k, v := range daemonSet.Labels {
-				toUpdate.Labels[k] = v
+func (r *ReconcileSubmariner) reconcileServiceDiscovery(submariner *submarinerv1alpha1.Submariner, reqLogger logr.Logger, isEnabled bool) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if isEnabled {
+			sd := newServiceDiscoveryCR(submariner)
```

Comment:
```suggestion
			var sd *submarinerv1alpha1.ServiceDiscovery
```

No need to initialize as it's overwritten on L219.